### PR TITLE
New high-level api module (QuickTx)

### PR DESCRIPTION
**Pull request by satran004**
Enhancement : https://github.com/bloxbean/cardano-client-lib/issues/163

This PR introduces new high-level api called "**QuickTx**". 

There is also a new interface "**TransactionEvaluator**" to calculate script cost.  
Currently we have only one implementation for this new interface. "Blockfrost" .  But the goal is to add more TransactionEvaluator implementations in future. For example: Aiken for offline tx evaluation.

**Status**: We have enough integration tests for the new apis. Unit tests will be added in next PR. 

It provides three main interfaces

1. **Tx** : For transaction from a regular account 
2. **ScriptTx** : For script transactions
3. **QuickTxBuilder** :  Transaction builder which composes multiple Tx or ScriptTx to build the final transaction

**Example:**

1. Simple Payment
```
        QuickTxBuilder quickTxBuilder = new QuickTxBuilder(backendService);
        Tx tx = new Tx()
                .payToAddress(receiver1, Amount.ada(1.5))
                .payToAddress(receiver2, Amount.ada(2.5))
                .attachMetadata(MessageMetadata.create().add("This is a test message"))
                .from(sender1Addr);

        Result<String> result = quickTxBuilder.compose(tx)
                .withSigner(SignerProviders.signerFrom(sender1))
                .complete();
```

2. Compose multiple regular transactions

```
       Tx tx1 = new Tx()
                .payToAddress(receiver1, Amount.ada(1.5))
                .payToAddress(receiver2, Amount.ada(2.5))
                .attachMetadata(MessageMetadata.create().add("This is a test message 2"))
                .from(sender1Addr);

        Tx tx2 = new Tx()
                .payToAddress(receiver3, Amount.ada(4.5))
                .from(sender2Addr);

        Result<String> result = quickTxBuilder
                .compose(tx1, tx2)
                .feePayer(sender1Addr)
                .withSigner(SignerProviders.signerFrom(sender1))
                .withSigner(SignerProviders.signerFrom(sender2))
                .completeAndWait(System.out::println);
```

3. Mint transaction

```
     Tx tx = new Tx()
                    .mintAssets(policy.getPolicyScript(), new Asset(assetName, qty), sender1.baseAddress())
                    .attachMetadata(MessageMetadata.create().add("Minting tx"))
                    .from(sender1.baseAddress());

      Result<String> result = quickTxBuilder.compose(tx)
                    .withSigner(SignerProviders.signerFrom(sender1))
                    .withSigner(SignerProviders.signerFrom(policy))
                    .complete();
```
4. A complex script transaction with multiple script invocations and mints

```
      ScriptTx scriptTx = new ScriptTx()
                .collectFrom(spendingUtxoOptional.get(), plutusData)
                .payToAddress(receiver2, Amount.lovelace(scriptAmt))
                .mintAsset(mintingScript, asset, PlutusData.unit(), sender2Addr)
                .mintAsset(plutusScript1, asset1, BigIntPlutusData.of(1), receiver1)
                .mintAsset(plutusScript2, asset2, BigIntPlutusData.of(2), sender1Addr)
                .mintAsset(plutusScript3, asset3, BigIntPlutusData.of(3), receiver1)
                .attachSpendingValidator(spendingScript)
                .withChangeAddress(sender2Addr);

        Result<String> result1 = quickTxBuilder.compose(scriptTx)
                .feePayer(sender2Addr)
                .withSigner(SignerProviders.signerFrom(sender1))
                .withSigner(SignerProviders.signerFrom(sender2))
                .mergeChangeOutputs(false)
                .withTxInspector(txn -> System.out.println(JsonUtil.getPrettyJson(txn)))
                .completeAndWait(System.out::println);
```

### CODE REVIEW

File: quicktx/src/main/java/com/bloxbean/cardano/client/quicktx/AbstractTx.java [hunk 1]
```diff
@@ -0,0 +1,453 @@
...
+     * @param amounts        List of Amount to send
+     * @param datumHash      datum hash
+     * @param datum          inline datum
+     * @param scriptRef      Reference Script
+     * @param scriptRefBytes Reference Script bytes
+     * @param mintOutput     If the asset in the output will be minted in this transaction, set this to true, otherwise false
+     * @param <T>
+     * @return T
+     */
+    protected <T> T payToAddress(String address, List<Amount> amounts, byte[] datumHash, PlutusData datum, Script scriptRef, byte[] scriptRefBytes, boolean mintOutput) {
```
[the remark this answers is not in the document]
Comment (satran004) [reply quicktx/src/main/java/com/bloxbean/cardano/client/quicktx/AbstractTx.java]: @sonatype-lift ignoreall

**Comment by matiwinnetou** [comment]
How does this work item play with the whole suppliers package idea?

File: core-api/src/main/java/com/bloxbean/cardano/client/api/TransactionEvaluator.java [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package com.bloxbean.cardano.client.api;
+
+import com.bloxbean.cardano.client.api.exception.ApiException;
+import com.bloxbean.cardano.client.api.model.EvaluationResult;
+import com.bloxbean.cardano.client.api.model.Result;
+
+import java.util.List;
+
+public interface TransactionEvaluator {
+    Result<List<EvaluationResult>> evaluateTx(byte[] cbor) throws ApiException;
```
Comment (matiwinnetou) [inline core-api/src/main/java/com/bloxbean/cardano/client/api/TransactionEvaluator.java:10]: 👍

**Comment by satran004** [comment]
> How does this work item play with the whole suppliers package idea?

@matiwinnetou 
```QuickTxBuilder``` internally depends only on supplier apis. For convenience, one of the constructor takes BackendService instance and extracts required suppliers from there.

But there is an another constructor with suppliers

```
QuickTxBuilder(UtxoSupplier utxoSupplier, ProtocolParamsSupplier protocolParamsSupplier,
                          TransactionProcessor transactionProcessor)
```

**Pull request merged**